File: determine_version_number.py
import re
import subprocess

lines = subprocess.check_output([
    "git",
    "log",
    "--date=iso8601-strict",
    "--pretty=format:%H %ad %d",
    "--decorate=full"
    ], encoding="UTF-8").split("\n")

lines = [line for line in lines if len(line.strip())]
tag = "initial"
since_last_tag = 0
for line in lines[::-1]:
    sha, dt, rest = line.split(" ", 2)
    match = re.search(r"\brefs/tags/(v\d[^),]+)", rest)
    if match:
        tag = match.group(1)
        since_last_tag = 0
    else:
        since_last_tag += 1
print(tag if since_last_tag == 0 else f"{tag}-dev{since_last_tag} ({sha[:7]})")

File: determine_version_number.mjs
const TAG_PREFIX = "refs/tags/"
const TAG_MATCH = /^refs\/tags\/v\d/
import {exec} from "child_process"
exec("git log  --date='format-local:%Y-%m-%dT%H:%M UTC' '--pretty=format:%H %ad %d' --decorate=full", {env: {TZ: "UTC0"}, HOME: ""}, 
  (error, stdout, stderr) => {
    if (error) {
      throw error;
      }
    const lines = stdout.split("\n").filter(l => l.length).reverse()
    const {tag, count, dt, sha} = lines.reduce(({tag, count}, line) => {
      const [sha, dt, ...rest] = line.split(" ")
      const tags = rest.filter(r => TAG_MATCH.exec(r)).map(
        r => r.slice(TAG_PREFIX.length, -1))
      if (tags.length === 0) {
        return {tag, count: count + 1, dt, sha}
      } else {
        return {tag: tags[0], count: 0, dt, sha}
      }
    }, {tag: "initial", count: 0, dt: "", sha: ""})
    exec("git status --porcelain=v1", {env: {HOME: ""}}, (error, stdout) => {
      if (error) {
        throw error;
      }
      const local_changes = stdout.split("\n").filter(l => l.length).length

      console.log(JSON.stringify({version: [
        `${tag}`,
        count ? `-dev+${count} (${sha.slice(0, 7)}) `: " ",
        local_changes ? `changes: ${local_changes} `: "",
        `(${dt.replace("T", " ")} UTC)`
      ].join("")}))
    })
  })
